### PR TITLE
fix: `noUncheckedIndexedAccess` issues

### LINKED ISSUE #1598: Usage with `noUncheckedIndexedAccess: true`
Hi! 👋 
      
I've enabled `noUncheckedIndexedAccess` in my ember app's `tsconfig.json`..

```json
"noUncheckedIndexedAccess": true,
```

..and I'm seeing the following error..

```
node_modules/ember-intl/addon/-private/utils/flatten.ts:27:9 - error TS2322: Type 'T | undefined' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'T | undefined'.

27         result[`${key}.${suffix}`] = hash[suffix];
           ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm unsure if this issue needs to be solved here, or if I can ignore this error with some extra config in `tsconfig.json`

I'm using `ember-intl@5.7.0` if that helps!

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/ember-intl/addon/-private/utils/flatten.ts b/node_modules/ember-intl/addon/-private/utils/flatten.ts
index d891fa9..16d9609 100644
--- a/node_modules/ember-intl/addon/-private/utils/flatten.ts
+++ b/node_modules/ember-intl/addon/-private/utils/flatten.ts
@@ -24,7 +24,10 @@ export default function flatten<T>(src: NestedStructure<T>): Record<string, T> {
       const hash = flatten(value);
 
       for (const suffix in hash) {
-        result[`${key}.${suffix}`] = hash[suffix];
+        const hashValue = hash[suffix];
+        if (typeof hashValue !== 'undefined') {
+          result[`${key}.${suffix}`] = hashValue;
+        }
       }
     } else {
       result[key] = value!;
```

<em>This issue body was [partially generated by patch-package](https://github.com/ds300/patch-package/issues/296).</em>

### PR DESCRIPTION
**Pull request by BarryThePenguin**
I fixed the `noUncheckedIndexedAccess` issues with the non-null assertion operator.

Fixing these issues in a different way would change the behaviour of the code, so I thought it would be helpful to get others input on alternative approaches.

Fixes #1598

### CODE REVIEW

File: addon/-private/formatters/format-relative.ts [hunk 1]
```diff
@@ -36,7 +36,7 @@ export default class FormatRelative extends Formatter<RelativeTimeFormatOptions>
     const { format } = formatOptions;
     let unit = formatOptions.unit;
     let opts: RelativeTimeFormatOptions = formatOptions;
-    if (!unit && format && intl.formats.relative && (opts = intl.formats.relative[format])) {
+    if (!unit && format && intl.formats.relative && (opts = intl.formats.relative[format]!)) {
```
Comment (BarryThePenguin) [inline addon/-private/formatters/format-relative.ts:39]: An alternative here would be to update line 39 to allow `undefined` options

```ts
let opts: RelativeTimeFormatOptions | undefined = formatOptions;
```

Which also matches the current behaviour

File: addon/-private/utils/flatten.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ export default function flatten<T>(src: NestedStructure<T>): Record<string, T> {
       const hash = flatten(value);
 
       for (const suffix in hash) {
-        result[`${key}.${suffix}`] = hash[suffix];
+        result[`${key}.${suffix}`] = hash[suffix]!;
```
Comment (BarryThePenguin) [inline addon/-private/utils/flatten.ts:27]: Here we could check that `hash[suffix]` is not `undefined`, but that would change which keys would be added to `result`.

We could update `result` to allow `undefined` values

```ts
const result = new EmptyObject() as Record<string, T | undefined>;
```

Line 30 has a similar issue

File: addon/macros/t.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ function partitionDynamicValuesAndStaticValues(options: Record<string, string |
     if (value instanceof Raw) {
       staticValues[key] = value.valueOf();
     } else {
-      dynamicValues[key] = value;
+      dynamicValues[key] = value!;
```
Comment (BarryThePenguin) [inline addon/macros/t.ts:18]: Here we could check that `value` is not `undefined`, but again, that would change which keys would be added to `dynamicValues`.

We could update `dynamicValues` to allow `undefined` values

```ts
const dynamicValues = new EmptyObject() as Record<string, string | undefined>;
```

File: tests/unit/formatters/format-relative-test.ts [hunk 4]
```diff
@@ -2,9 +2,9 @@ import { module, test } from 'qunit';
 import FormatRelative from 'ember-intl/-private/formatters/format-relative';
 import { createIntl, OnErrorFn } from '@formatjs/intl';
 
-function getIntl(locales: string | string[]) {
+function getIntl(locale: string) {
   return createIntl({
-    locale: Array.isArray(locales) ? locales[0] : locales,
+    locale: locale,
```
Comment (BarryThePenguin) [inline tests/unit/formatters/format-relative-test.ts:7]: `locales` was never an array, so I just changed this to a `string`

File: addon/-private/formatters/format-relative.ts (hunk 1 above)
Comment (longlho) [reply addon/-private/formatters/format-relative.ts:39]: yeah that's actually better

File: addon/-private/utils/flatten.ts (hunk 2 above)
Comment (longlho) [reply addon/-private/utils/flatten.ts:27]: we can check that `hash[suffix]` is not `undefined` as well

File: addon/macros/t.ts (hunk 3 above)
Comment (longlho) [reply addon/macros/t.ts:18]: let's exclude `undefined`

**Review by longlho — CHANGES_REQUESTED** [review]

**Review by longlho — APPROVED** [review]

File: tests/unit/formatters/format-relative-test.ts (hunk 4 above)
Comment (MrChocolatine) [reply tests/unit/formatters/format-relative-test.ts:7]: Not essential but

```suggestion
    locale,
```

**Comment by longlho** [comment]
Thanks a lot for your contributions!

**Pull request merged**